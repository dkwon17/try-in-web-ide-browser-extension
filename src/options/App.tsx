/*-----------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Licensed under the MIT License. See LICENSE file in the project root for license information.
 *-----------------------------------------------------------------------------------------------*/

import { useEffect, useState } from "react";

import "@patternfly/react-core/dist/styles/base.css";
import "@patternfly/patternfly/utilities/Spacing/spacing.css";
import "@patternfly/patternfly/utilities/Float/float.css";

import {
    Button,
    Card,
    CardBody,
    Divider,
    Split,
    SplitItem,
    Form,
    FormGroup,
    TextInput,
    CardTitle,
} from "@patternfly/react-core";
import { ExclamationCircleIcon } from "@patternfly/react-icons";
import {
    Endpoint,
    getEndpoints,
    saveEndpoints,
} from "../preferences/preferences";
import { EndpointsList } from "./EndpointsList";
import "./styles/App.css";

export const App = () => {
    const [endpoints, setEndpoints] = useState<Endpoint[]>([]);
    const [newEndpointUrl, setNewEndpointUrl] = useState<string>("");
    
    type validate = "success" | "error" | "default";
    const [newEndpointStatus, setNewEndpointStatus] = useState<validate>("default");

    useEffect(() => {
        const fetchData = async () => {
            const endpoints = await getEndpoints();
            setEndpoints(endpoints);
        };
        fetchData().catch(console.error);
    }, []);

    const handleNewEndpointUrlChange = (newUrl: string, _event: React.FormEvent<HTMLInputElement>) => {
        setNewEndpointUrl(newUrl);
        if (newUrl === "") {
          setNewEndpointStatus("default");
        } else if (isUrl(newUrl)) {
          setNewEndpointStatus("success");
        } else {
          setNewEndpointStatus("error");
        }
    };

    const isUrl = (str: string) => {
        try {
            new URL(str);
        } catch {
            return false;
        }
        return true;
    }

    const addNewEndpoint = async () => {
        const sanitizedEndpoint = sanitizeEndpoint(newEndpointUrl);
        const newEndpoints = endpoints.concat({
            url: sanitizedEndpoint,
            active: false,
            readonly: false,
        });
        await saveEndpoints(newEndpoints);
        await setEndpoints(newEndpoints);
        setNewEndpointUrl("");
        setNewEndpointStatus("default");
    };

    const sanitizeEndpoint = (str: string) => {
        let res = str;
        while (res.charAt(res.length - 1) === "/") {
            res = res.substring(0, res.length - 1);
        }
        return res;
    }

    const setDefault = async (endpoint: Endpoint) => {
        const newEndpoints = [...endpoints];
        newEndpoints.forEach((e) => {
            e.active = e == endpoint;
        });
        await saveEndpoints(newEndpoints);
        await setEndpoints(newEndpoints);
    };

    const deleteEndpoint = async (endpoint: Endpoint) => {
        const newEndpoints = endpoints.filter((e) => e != endpoint);
        await saveEndpoints(newEndpoints);
        await setEndpoints(newEndpoints);
    };

    const list = endpoints.length && (
        <EndpointsList
            endpoints={endpoints}
            onClickSetDefault={setDefault}
            onClickDelete={deleteEndpoint}
        />
    );

    return (
        <Card>
            <CardTitle>Dev Spaces endpoints</CardTitle>
            <CardBody>
                {list}
                <Divider className="pf-u-mt-md pf-u-mb-md" />
                <Split>
                    <SplitItem className="form-text-input">
                        <Form>
                            <FormGroup
                                validated={newEndpointStatus}
                                helperTextInvalid="The new endpoint must be a URL"
                                helperTextInvalidIcon={<ExclamationCircleIcon />}>
                                <TextInput
                                    type="text"
                                    aria-label="new endpoint"
                                    validated={newEndpointStatus}
                                    value={newEndpointUrl}
                                    placeholder="Add endpoint"
                                    onChange={handleNewEndpointUrlChange}
                                />
                            </FormGroup>
                        </Form>
                    </SplitItem>
                    <SplitItem className="form-fill" isFilled></SplitItem>
                    <SplitItem>
                        <Button variant="primary" onClick={addNewEndpoint} isDisabled={newEndpointUrl.length === 0 || newEndpointStatus === "error"}>
                            Add
                        </Button>
                    </SplitItem>
                </Split>
            </CardBody>
        </Card>
    );
};
