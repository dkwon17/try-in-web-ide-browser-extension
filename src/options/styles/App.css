.pf-c-card {
    --pf-c-card__title--FontWeight: var(--pf-global--FontWeight--normal) !important;
    --pf-c-card__title--not--last-child--PaddingBottom: var(--pf-global--spacer--sm) !important;
    --pf-c-card__title--FontSize: var(--pf-global--FontSize--3xl) !important;
    --pf-c-card--first-child--PaddingTop: var(--pf-global--spacer--md) !important;
    --pf-c-card--child--PaddingBottom: var(--pf-global--spacer--xl) !important;
    --pf-c-card--child--PaddingRight: var(--pf-global--spacer--xl) !important;
    --pf-c-card--child--PaddingLeft: var(--pf-global--spacer--xl) !important;
    height: 100%;
}

/* styling the list of hostnames */
.pf-c-simple-list {
    --pf-c-simple-list__item-link--PaddingLeft: 0 !important;

    /* disable hover background change */
    --pf-c-simple-list__item-link--hover--BackgroundColor: var(--pf-global--BackgroundColor--light-100) !important;
    --pf-c-simple-list__item-link--Color: var(--pf-global--Color--100) !important;

    /* disable styles when list item is clicked on */
    --pf-c-simple-list__item-link--m-current--BackgroundColor: var(--pf-global--BackgroundColor--light-100) !important;
    --pf-c-simple-list__item-link--m-current--Color: var(--pf-global--Color--100) !important;
    --pf-c-simple-list__item-link--active--BackgroundColor: var(--pf-global--BackgroundColor--light-100) !important;
    --pf-c-simple-list__item-link--active--Color: var(--pf-global--Color--100) !important;
    --pf-c-simple-list__item-link--FontSize: var(--pf-global--FontSize--md) !important;
}

/* styling the dropdown menu button */
.pf-c-dropdown {
    --pf-c-dropdown__toggle--PaddingTop: 0 !important;
    --pf-c-dropdown__toggle--PaddingRight: 0 !important;
    --pf-c-dropdown__toggle--PaddingBottom: 0 !important;
    --pf-c-dropdown__toggle--PaddingLeft: 0 !important;
    --pf-c-dropdown__toggle--before--BorderWidth: 0 !important;
    --pf-c-dropdown--m-expanded__toggle--before--BorderBottomWidth: 0 !important;
    --pf-c-dropdown__toggle--focus--before--BorderBottomWidth: 0 !important;
    --pf-c-dropdown__toggle--active--before--BorderBottomWidth: 0 !important;
    --pf-c-dropdown__toggle--Color: var(--pf-global--palette--black-600) !important;
}

.pf-c-simple-list__item-link {
    cursor: default;
}

.form-text-input {
    width: 100%;
}

.form-fill {
    width: var(--pf-global--spacer--sm);
}

