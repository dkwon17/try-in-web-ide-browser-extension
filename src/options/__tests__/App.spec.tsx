/*-----------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Licensed under the MIT License. See LICENSE file in the project root for license information.
 *-----------------------------------------------------------------------------------------------*/

import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

import { App } from "../App";

const preferencesMock = require("../../preferences/preferences");

jest.mock("../../preferences/preferences");

it("renders correctly with one endpoint", async () => {
    preferencesMock.setEndpoints([
        { url: "https://url-1.com", active: true, readonly: true },
    ]);
    const { findByText, asFragment } = render(<App />);
    await findByText("https://url-1.com");
    expect(asFragment()).toMatchSnapshot();
});

it("renders correctly with multiple endpoints", async () => {
    preferencesMock.setEndpoints([
        { url: "https://url-1.com", active: true, readonly: true },
        { url: "https://url-2.com", active: false, readonly: false },
        { url: "https://url-3.com", active: false, readonly: false },
    ]);
    const { findByText, asFragment } = render(<App />);
    await findByText("https://url-1.com");
    await findByText("https://url-2.com");
    await findByText("https://url-3.com");
    expect(asFragment()).toMatchSnapshot();
});

it("should have the same number of endpoints as links in the endpoints list (1)", async () => {
    preferencesMock.setEndpoints([
        { url: "https://url-1.com", active: true, readonly: true },
    ]);
    const { findByText, findByTestId } = render(<App />);
    await findByText("https://url-1.com");

    const list = await findByTestId("endpoints-list");
    expect(list.querySelectorAll("li").length).toEqual(1);
});

it("should have the same number of endpoints as links in the endpoints list (3)", async () => {
    preferencesMock.setEndpoints([
        { url: "https://url-1.com", active: true, readonly: true },
        { url: "https://url-2.com", active: false, readonly: false },
        { url: "https://url-3.com", active: false, readonly: false },
    ]);
    const { findByText, findByTestId } = render(<App />);
    await findByText("https://url-1.com");
    await findByText("https://url-2.com");
    await findByText("https://url-3.com");

    const list = await findByTestId("endpoints-list");
    expect(list.querySelectorAll("li").length).toEqual(3);
});

it("has same number of endpoints as links in the endpoints list (3)", async () => {
    preferencesMock.setEndpoints([
        { url: "https://url-1.com", active: true, readonly: true },
        { url: "https://url-2.com", active: false, readonly: false },
        { url: "https://url-3.com", active: false, readonly: false },
    ]);
    const { findByText, findByTestId } = render(<App />);
    await findByText("https://url-1.com");
    await findByText("https://url-2.com");
    await findByText("https://url-3.com");

    const list = await findByTestId("endpoints-list");
    expect(list.querySelectorAll("li")).toHaveLength(3);
});

it("should have only one default label", async () => {
    preferencesMock.setEndpoints([
        { url: "https://url-1.com", active: true, readonly: true },
        { url: "https://url-2.com", active: false, readonly: false },
        { url: "https://url-3.com", active: false, readonly: false },
    ]);
    const { findAllByText } = render(<App />);
    const items = await findAllByText("Default");
    expect(items).toHaveLength(1);
});
