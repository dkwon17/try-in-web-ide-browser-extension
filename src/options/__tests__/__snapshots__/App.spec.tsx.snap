// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with multiple endpoints 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-2"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      Dev Spaces endpoints
    </div>
    <div
      class="pf-c-card__body"
    >
      <div
        class="pf-c-simple-list"
        data-testid="endpoints-list"
      >
        <ul>
          <li
            class=""
          >
            <a
              class="pf-c-simple-list__item-link"
              href=""
              tabindex="0"
            >
              <div
                class="pf-l-split"
              >
                <div
                  class="pf-l-split__item"
                >
                  <span
                    class="pf-c-truncate"
                  >
                    <span
                      class="pf-c-truncate__start"
                    >
                      https://url-1.com
                    </span>
                  </span>
                </div>
                <div
                  class="pf-l-split__item"
                >
                  <span
                    aria-label="Default endpoint"
                    class="pf-c-label pf-m-blue pf-m-compact pf-u-ml-sm pf-u-mr-sm"
                  >
                    <span
                      class="pf-c-label__content"
                    >
                      Default
                    </span>
                  </span>
                </div>
                <div
                  class="pf-l-split__item pf-m-fill"
                />
                <div
                  class="pf-l-split__item"
                >
                  <div
                    aria-label="Toggle dropdown"
                    class="pf-c-dropdown pf-m-align-right pf-u-float-right"
                    data-ouia-component-id="OUIA-Generated-Dropdown-2"
                    data-ouia-component-type="PF4/Dropdown"
                    data-ouia-safe="true"
                    title="Toggle dropdown"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-label="Actions"
                      class="pf-c-dropdown__toggle"
                      data-ouia-component-id="OUIA-Generated-DropdownToggle-2"
                      data-ouia-component-type="PF4/DropdownToggle"
                      data-ouia-safe="true"
                      id="pf-dropdown-toggle-id-1"
                      type="button"
                    >
                      <span>
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 192 512"
                          width="1em"
                        >
                          <path
                            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="pf-c-simple-list__item-link"
              href=""
              tabindex="0"
            >
              <div
                class="pf-l-split"
              >
                <div
                  class="pf-l-split__item"
                >
                  <span
                    class="pf-c-truncate"
                  >
                    <span
                      class="pf-c-truncate__start"
                    >
                      https://url-2.com
                    </span>
                  </span>
                </div>
                <div
                  class="pf-l-split__item pf-m-fill"
                />
                <div
                  class="pf-l-split__item"
                >
                  <div
                    aria-label="Toggle dropdown"
                    class="pf-c-dropdown pf-m-align-right pf-u-float-right"
                    data-ouia-component-id="OUIA-Generated-Dropdown-3"
                    data-ouia-component-type="PF4/Dropdown"
                    data-ouia-safe="true"
                    title="Toggle dropdown"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-label="Actions"
                      class="pf-c-dropdown__toggle"
                      data-ouia-component-id="OUIA-Generated-DropdownToggle-3"
                      data-ouia-component-type="PF4/DropdownToggle"
                      data-ouia-safe="true"
                      id="pf-dropdown-toggle-id-2"
                      type="button"
                    >
                      <span>
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 192 512"
                          width="1em"
                        >
                          <path
                            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </a>
          </li>
          <li
            class=""
          >
            <a
              class="pf-c-simple-list__item-link"
              href=""
              tabindex="0"
            >
              <div
                class="pf-l-split"
              >
                <div
                  class="pf-l-split__item"
                >
                  <span
                    class="pf-c-truncate"
                  >
                    <span
                      class="pf-c-truncate__start"
                    >
                      https://url-3.com
                    </span>
                  </span>
                </div>
                <div
                  class="pf-l-split__item pf-m-fill"
                />
                <div
                  class="pf-l-split__item"
                >
                  <div
                    aria-label="Toggle dropdown"
                    class="pf-c-dropdown pf-m-align-right pf-u-float-right"
                    data-ouia-component-id="OUIA-Generated-Dropdown-4"
                    data-ouia-component-type="PF4/Dropdown"
                    data-ouia-safe="true"
                    title="Toggle dropdown"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-label="Actions"
                      class="pf-c-dropdown__toggle"
                      data-ouia-component-id="OUIA-Generated-DropdownToggle-4"
                      data-ouia-component-type="PF4/DropdownToggle"
                      data-ouia-safe="true"
                      id="pf-dropdown-toggle-id-3"
                      type="button"
                    >
                      <span>
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 192 512"
                          width="1em"
                        >
                          <path
                            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </a>
          </li>
        </ul>
      </div>
      <hr
        class="pf-c-divider pf-u-mt-md pf-u-mb-md"
      />
      <div
        class="pf-l-split"
      >
        <div
          class="pf-l-split__item form-text-input"
        >
          <form
            class="pf-c-form"
            novalidate=""
          >
            <div
              class="pf-c-form__group"
            >
              <div
                class="pf-c-form__group-control"
              >
                <input
                  aria-invalid="false"
                  aria-label="new endpoint"
                  class="pf-c-form-control"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                  data-ouia-component-type="PF4/TextInput"
                  data-ouia-safe="true"
                  placeholder="Add endpoint"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </form>
        </div>
        <div
          class="pf-l-split__item pf-m-fill form-fill"
        />
        <div
          class="pf-l-split__item"
        >
          <button
            aria-disabled="true"
            class="pf-c-button pf-m-primary pf-m-disabled"
            data-ouia-component-id="OUIA-Generated-Button-primary-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            disabled=""
            type="button"
          >
            Add
          </button>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`renders correctly with one endpoint 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      Dev Spaces endpoints
    </div>
    <div
      class="pf-c-card__body"
    >
      <div
        class="pf-c-simple-list"
        data-testid="endpoints-list"
      >
        <ul>
          <li
            class=""
          >
            <a
              class="pf-c-simple-list__item-link"
              href=""
              tabindex="0"
            >
              <div
                class="pf-l-split"
              >
                <div
                  class="pf-l-split__item"
                >
                  <span
                    class="pf-c-truncate"
                  >
                    <span
                      class="pf-c-truncate__start"
                    >
                      https://url-1.com
                    </span>
                  </span>
                </div>
                <div
                  class="pf-l-split__item"
                >
                  <span
                    aria-label="Default endpoint"
                    class="pf-c-label pf-m-blue pf-m-compact pf-u-ml-sm pf-u-mr-sm"
                  >
                    <span
                      class="pf-c-label__content"
                    >
                      Default
                    </span>
                  </span>
                </div>
                <div
                  class="pf-l-split__item pf-m-fill"
                />
                <div
                  class="pf-l-split__item"
                >
                  <div
                    aria-label="Toggle dropdown"
                    class="pf-c-dropdown pf-m-align-right pf-u-float-right"
                    data-ouia-component-id="OUIA-Generated-Dropdown-1"
                    data-ouia-component-type="PF4/Dropdown"
                    data-ouia-safe="true"
                    title="Toggle dropdown"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-label="Actions"
                      class="pf-c-dropdown__toggle"
                      data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
                      data-ouia-component-type="PF4/DropdownToggle"
                      data-ouia-safe="true"
                      id="pf-dropdown-toggle-id-0"
                      type="button"
                    >
                      <span>
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 192 512"
                          width="1em"
                        >
                          <path
                            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </a>
          </li>
        </ul>
      </div>
      <hr
        class="pf-c-divider pf-u-mt-md pf-u-mb-md"
      />
      <div
        class="pf-l-split"
      >
        <div
          class="pf-l-split__item form-text-input"
        >
          <form
            class="pf-c-form"
            novalidate=""
          >
            <div
              class="pf-c-form__group"
            >
              <div
                class="pf-c-form__group-control"
              >
                <input
                  aria-invalid="false"
                  aria-label="new endpoint"
                  class="pf-c-form-control"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                  data-ouia-component-type="PF4/TextInput"
                  data-ouia-safe="true"
                  placeholder="Add endpoint"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </form>
        </div>
        <div
          class="pf-l-split__item pf-m-fill form-fill"
        />
        <div
          class="pf-l-split__item"
        >
          <button
            aria-disabled="true"
            class="pf-c-button pf-m-primary pf-m-disabled"
            data-ouia-component-id="OUIA-Generated-Button-primary-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            disabled=""
            type="button"
          >
            Add
          </button>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
